operator-wg/whitepaper-community-review/README.md

```diff
@@ -0,0 +1,882 @@
+
+### Further reading
+There are lots of more best practices like:
+* An operator shouldn't install other operators
+* Operators shouldn't make assumptions about the namespaces they are deployed in, but also
+* Use an SDK for writing operators
+
+and many other best practices might be found on the internet. More of them could be found on following sources:
+* https://github.com/operator-framework/community-operators/blob/master/docs/best-practices.md
+* https://cloud.google.com/blog/products/containers-kubernetes/best-practices-for-building-kubernetes-operators-and-stateful-apps
+    
+## Designing Operators
+
+The previous chapter describes a use case for an operator that was one of the first operators ever. With no claim
+of completeness, this chapter deals with some best practice when writing
+own Operators, based on our own experience or described by the
+community. However, without clear knowledge of the actual state and
+without clear ideas of what we want to achieve, we also need some
+methods and techniques specifying what our Operator should do.
+Therefore, we will also have to deal with some aspects of requirement
+engineering.
+
+### Requirement analysis
+
+A key promise of Kubernetes is that it enables the automation of operational
+tasks to deploy, scale, and manage containerized applications across
+multiple environments with no (or minimal) human intervention. In
+Kubernetes, stateless cloud-native applications are well suited for
+horizontal scaling, automated self-healing restarts, or progressive
+rollout of new containers. However, stateful applications with complex
+components running in clustered or distributed environments are not
+always well suited for this type of container-based infrastructure. They
+still require human interaction when it comes to persistence, upgrades,
+or high availability to remain in a stable state.
+
+True, Kubernetes solves these issues in a novel way by creating and
+managing custom applications using Operators. However, and here is the
+first question: as a developer, do you really know how this type of
+application works and interacts both internally and externally? How do
+the day-to-day IT operations work? How is the application backed up
+(including recovery)? What steps are necessary in case of failovers or
+outages, are there any dependencies between the software components?
+
+It is therefore strongly recommended that a comprehensive requirement
+analysis is needed to determine the requirements or conditions of an
+Operator. Requirement analysis is critical to the success or failure of
+Operators. All requirements should be documented, measurable, testable,
+traceable, related to identified requirements, and defined at a level of
+detail sufficient for system design.
+
+Best practice:
+
+1.  Do not assume to know what is needed.
```

Comment:
I am not sure this fits as the opener of best practice. 
There are circumstances an engineer with their colleagues might know exactly what is required. What is the point behind this statement?

Reply:
Hi, thanks for your comment. and you are right, I struggled with that statement too. However, I decided to keep this statement as it is about Requirement Engineering (RE) rather than an Operator. Assumptions are reasonably believed to be true based on experience, e.g. lessons learned.  Engineers with a lot of experience can  make assumptions much easier, but engineers with less experience should therefore not make assumptions, but first use techniques and methods of RE to gain the necessary experience.

Reply:
Would mind making the Requirement Engineering focus more specific in this statement? Just so it doesn't seem too random.